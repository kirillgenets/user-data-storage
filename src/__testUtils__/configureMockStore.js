import configureStore from 'redux-mock-store';

const configureMockStore = () => configureStore([]);

export default configureMockStore;
