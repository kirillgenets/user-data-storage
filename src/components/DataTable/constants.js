import { FieldName } from '../../constants';

export const COLUMNS = [
  {
    key: FieldName.FirstName,
    name: 'First name',
  },
  {
    key: FieldName.LastName,
    name: 'Last name',
  },
  {
    key: FieldName.Email,
    name: 'E-Mail',
  },
  {
    key: FieldName.BirthDate,
    name: 'Date of birth',
  },
];

export const ITEMS_PER_PAGE = 5;
