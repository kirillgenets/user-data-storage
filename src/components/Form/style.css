.form {
  display: flex;
  flex-direction: column;
  gap: 15px;

  width: 300px;
  margin: 0 auto;
}

.form__field {
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.form__input {
  padding: 8px;
}

.form__label {
  font-weight: 600;
}

.form__error {
  margin: 0;

  color: #ad4e4e;
}

.form__submit {
  margin: 0 auto;
  padding: 10px 20px;

  cursor: pointer;
}
