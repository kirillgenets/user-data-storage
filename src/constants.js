export const FieldName = {
  FirstName: 'firstName',
  LastName: 'lastName',
  Email: 'email',
  BirthDate: 'birthDate',
};
