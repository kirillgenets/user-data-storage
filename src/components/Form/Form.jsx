import { useCallback, useRef, useState } from 'react';
import { useDispatch } from 'react-redux';
import Validator from 'simple-react-validator';

import { addCustomer } from '../../actions/customers';
import { FieldName } from '../../constants';

import './style.css';

const VALIDATION_RULES = {
  [FieldName.FirstName]: 'required|alpha_space',
  [FieldName.LastName]: 'required|alpha_space',
  [FieldName.Email]: 'required|email',
  [FieldName.BirthDate]: 'required|regex:[0-9]{4}-[0-9]{2}-[0-9]{2}',
};

const DEFAULT_FORM_DATA = {
  [FieldName.FirstName]: '',
  [FieldName.LastName]: '',
  [FieldName.Email]: '',
  [FieldName.BirthDate]: '',
};

const Form = () => {
  const validatorRef = useRef(
    new Validator({
      element: (message) => <p className="form__error">{message}</p>,
    }),
  );
  const [formData, setFormData] = useState(DEFAULT_FORM_DATA);
  const [, setIsError] = useState(false);

  const dispatch = useDispatch();

  const handleFieldChange = useCallback(
    (name) => (evt) => {
      setFormData((prevFormData) => ({ ...prevFormData, [name]: evt.target.value }));
    },
    [],
  );

  const handleSubmit = useCallback(
    (evt) => {
      evt.preventDefault();
      if (validatorRef.current.allValid()) {
        dispatch(addCustomer(formData));
        setFormData(DEFAULT_FORM_DATA);
        validatorRef.current.hideMessages();
        setIsError(false);
      } else {
        validatorRef.current.showMessages();
        setIsError(true);
      }
    },
    [formData, dispatch],
  );

  return (
    <form action="#" className="form" onSubmit={handleSubmit}>
      <div className="form__field">
        <label htmlFor={FieldName.FirstName} className="form__label">
          First name:
        </label>
        <input
          type="text"
          name={FieldName.FirstName}
          id={FieldName.FirstName}
          value={formData[FieldName.FirstName]}
          className="form__input"
          placeholder="Please, type your first name..."
          onChange={handleFieldChange(FieldName.FirstName)}
        />
        {validatorRef.current.message(
          FieldName.FirstName,
          formData[FieldName.FirstName],
          VALIDATION_RULES[FieldName.FirstName],
        )}
      </div>
      <div className="form__field">
        <label htmlFor={FieldName.LastName} className="form__label">
          Last name:
        </label>
        <input
          type="text"
          name={FieldName.LastName}
          id={FieldName.LastName}
          value={formData[FieldName.LastName]}
          className="form__input"
          placeholder="Please, type your last name..."
          onChange={handleFieldChange(FieldName.LastName)}
        />
        {validatorRef.current.message(
          FieldName.LastName,
          formData[FieldName.LastName],
          VALIDATION_RULES[FieldName.LastName],
        )}
      </div>
      <div className="form__field">
        <label htmlFor={FieldName.Email} className="form__label">
          Email:
        </label>
        <input
          type="email"
          name={FieldName.Email}
          id={FieldName.Email}
          value={formData[FieldName.Email]}
          className="form__input"
          placeholder="Please, type your email..."
          onChange={handleFieldChange(FieldName.Email)}
        />
        {validatorRef.current.message(
          FieldName.Email,
          formData[FieldName.Email],
          VALIDATION_RULES[FieldName.Email],
        )}
      </div>
      <div className="form__field">
        <label htmlFor={FieldName.BirthDate} className="form__label">
          Date of birth:
        </label>
        <input
          type="date"
          name={FieldName.BirthDate}
          id={FieldName.BirthDate}
          value={formData[FieldName.BirthDate]}
          className="form__input"
          placeholder="Please, type your date of birth..."
          onChange={handleFieldChange(FieldName.BirthDate)}
        />
        {validatorRef.current.message(
          FieldName.BirthDate,
          formData[FieldName.BirthDate],
          VALIDATION_RULES[FieldName.BirthDate],
        )}
      </div>
      <button type="submit" className="form__submit">
        Submit
      </button>
    </form>
  );
};

export default Form;
