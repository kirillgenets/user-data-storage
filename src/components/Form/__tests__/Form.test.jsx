import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { FieldName } from '../../../constants';

import configureMockStore from '../../../__testUtils__/configureMockStore';

import Form from '../Form';

const mockStore = configureMockStore();

describe('Form', () => {
  it('should render correctly', () => {
    const store = mockStore({});
    const component = mount(
      <Provider store={store}>
        <Form />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
    expect(component.find('.form')).toHaveLength(1);
  });

  it('should validate first name', () => {
    const store = mockStore({});
    const component = mount(
      <Provider store={store}>
        <Form validateOnChange />
      </Provider>,
    );

    component
      .find(`.form__input[name="${FieldName.FirstName}"]`)
      .simulate('change', { target: { value: '123' } });

    component
      .find(`.form__input[name="${FieldName.LastName}"]`)
      .simulate('change', { target: { value: 'test' } });

    component
      .find(`.form__input[name="${FieldName.Email}"]`)
      .simulate('change', { target: { value: 'test@example.com' } });

    component
      .find(`.form__input[name="${FieldName.BirthDate}"]`)
      .simulate('change', { target: { value: '1995-01-01' } });

    expect(component.find('.form__error')).toHaveLength(1);
    expect(component.find('.form__field:first-child .form__error')).toHaveLength(1);
  });

  it('should validate last name', () => {
    const store = mockStore({});
    const component = mount(
      <Provider store={store}>
        <Form validateOnChange />
      </Provider>,
    );

    component
      .find(`.form__input[name="${FieldName.FirstName}"]`)
      .simulate('change', { target: { value: 'test' } });

    component
      .find(`.form__input[name="${FieldName.LastName}"]`)
      .simulate('change', { target: { value: '123' } });

    component
      .find(`.form__input[name="${FieldName.Email}"]`)
      .simulate('change', { target: { value: 'test@example.com' } });

    component
      .find(`.form__input[name="${FieldName.BirthDate}"]`)
      .simulate('change', { target: { value: '1995-01-01' } });

    expect(component.find('.form__error')).toHaveLength(1);
    expect(component.find('form').childAt(1).find('.form__error')).toHaveLength(1);
  });

  it('should validate email', () => {
    const store = mockStore({});
    const component = mount(
      <Provider store={store}>
        <Form validateOnChange />
      </Provider>,
    );

    component
      .find(`.form__input[name="${FieldName.FirstName}"]`)
      .simulate('change', { target: { value: 'test' } });

    component
      .find(`.form__input[name="${FieldName.LastName}"]`)
      .simulate('change', { target: { value: 'test' } });

    component
      .find(`.form__input[name="${FieldName.Email}"]`)
      .simulate('change', { target: { value: '123' } });

    component
      .find(`.form__input[name="${FieldName.BirthDate}"]`)
      .simulate('change', { target: { value: '1995-01-01' } });

    expect(component.find('.form__error')).toHaveLength(1);
    expect(component.find('form').childAt(2).find('.form__error')).toHaveLength(1);
  });

  it('should validate birthdate', () => {
    const store = mockStore({});
    const component = mount(
      <Provider store={store}>
        <Form validateOnChange />
      </Provider>,
    );

    component
      .find(`.form__input[name="${FieldName.FirstName}"]`)
      .simulate('change', { target: { value: 'test' } });

    component
      .find(`.form__input[name="${FieldName.LastName}"]`)
      .simulate('change', { target: { value: 'test' } });

    component
      .find(`.form__input[name="${FieldName.Email}"]`)
      .simulate('change', { target: { value: 'test@example.com' } });

    component
      .find(`.form__input[name="${FieldName.BirthDate}"]`)
      .simulate('change', { target: { value: '123' } });

    expect(component.find('.form__error')).toHaveLength(1);
    expect(component.find('form').childAt(3).find('.form__error')).toHaveLength(1);
  });
});
