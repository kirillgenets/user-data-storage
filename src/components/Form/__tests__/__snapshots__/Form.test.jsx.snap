// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
<form
  action="#"
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form__field"
  >
    <label
      className="form__label"
      htmlFor="firstName"
    >
      First name:
    </label>
    <input
      className="form__input"
      id="firstName"
      name="firstName"
      onChange={[Function]}
      placeholder="Please, type your first name..."
      type="text"
      value=""
    />
  </div>
  <div
    className="form__field"
  >
    <label
      className="form__label"
      htmlFor="lastName"
    >
      Last name:
    </label>
    <input
      className="form__input"
      id="lastName"
      name="lastName"
      onChange={[Function]}
      placeholder="Please, type your last name..."
      type="text"
      value=""
    />
  </div>
  <div
    className="form__field"
  >
    <label
      className="form__label"
      htmlFor="email"
    >
      Email:
    </label>
    <input
      className="form__input"
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Please, type your email..."
      type="email"
      value=""
    />
  </div>
  <div
    className="form__field"
  >
    <label
      className="form__label"
      htmlFor="birthDate"
    >
      Date of birth:
    </label>
    <input
      className="form__input"
      id="birthDate"
      name="birthDate"
      onChange={[Function]}
      placeholder="Please, type your date of birth..."
      type="date"
      value=""
    />
  </div>
  <button
    className="form__submit"
    type="submit"
  >
    Submit
  </button>
</form>
`;
